[sparsity] Sparsifier class

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* **#58704 [sparsity] Sparsifier class**
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

Commit:
[sparse] Sparsifier class

[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"




[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"




[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"




[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"




[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"




[ghstack-poisoned]

Commit:
Update on "[sparse] Sparsifier class"


Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparse] Sparsifier class"


Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

raghuramank100 commented on test/ao/test_sparse_kernels.py:
```diff
@@ -156,6 +156,7 @@ def test_sparse_qlinear(self):
                 model.qconfig = tq.get_default_qconfig('fbgemm')
                 qmodel = copy.deepcopy(model)
                 sqmodel = copy.deepcopy(model)
+                sqmodel.linear.mask = mask
```
NIT: mask would be an attribute of the fake-sparse module

Commit:
Update on "[sparse] Sparsifier class"


Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

raghuramank100 commented on test/ao/sparsity/test_kernels.py:
```diff
@@ -156,6 +156,7 @@ def test_sparse_qlinear(self):
                 model.qconfig = tq.get_default_qconfig('fbgemm')
                 qmodel = copy.deepcopy(model)
                 sqmodel = copy.deepcopy(model)
+                sqmodel.linear.mask = mask
```
Not sure what happened to previous comments: This should be sqmodel.linear.fake_sparse.mask

z-a-f replied:
That's not e en supposed to be tested here -- this bled through from some other PR

raghuramank100 commented on test/run_test.py:
```diff
@@ -33,6 +33,7 @@
     'test_import_time',
     'test_public_bindings',
     'test_type_hints',
+    'test_ao',
```
Should this be test_ao_sparsity instead?

z-a-f replied:
This is weird -- we changed it with @kazhou, weirdly it is still here.

raghuramank100 commented on torch/ao/sparsity/_variables.py:
```diff
@@ -0,0 +1,25 @@
+import torch
```
Suggest calling this sparsity_mappings.py, similar to: 
https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantization_mappings.py

raghuramank100 commented on torch/ao/sparsity/_variables.py:
```diff
@@ -0,0 +1,25 @@
+import torch
+
+def get_sparse_mapping():
```
Can these be plain dicts instead of functions, to align with quantization (see https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantization_mappings.py#L72)

z-a-f replied:
The quantization_mappings also use functions. In fact, using regular dicts is a bad idea, as they are mutable

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+    nn.Linear
+}
+
+def _module_to_path(model, layer, prefix=''):
```
NIT: Suggest calling this _module_to_fqn instead.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+            return child_path
+    return None
+
+def _path_to_module(model, path):
```
Do both these functions work for sequentials? Also, for consistency suggest calling this _fqn_to_module

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+        if self.config is None:
+            # Add all models to the config
+            self.config = []
+            stack = [model]
```
Why do we need a while loop here? Cant we just check for this with a for loop followed by an if?.

Also, in this case, we should have the user specify the list of modules, instead of passing the entire model

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+            local_args['path'] = module_path
+            self.module_groups.append(local_args)
+
+    def __getstate__(self):
```
Do we intend for sparsifier objects to be pickleable? module groups contain references to modules, so they shouldnt be pickleable

z-a-f replied:
At this stage the module_groups have only the FQNs, which we want to pickle, right?

raghuramank100 replied:
Yes, that is correct.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+        """
+        for config in self.module_groups:
+            module = config['module']
+            if not getattr(module, 'mask', None):
```
Lets replace this with fakesparse module insertions, which should contain the mask

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+                module_config = {'module': module_config}
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
```
module also needs to a reference to the weights so that step can use it.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,167 @@
+        self.module_groups = []
+        self.enable_mask_update = False
+
+        self.model = model
```
Can we eliminate the need to have a model as an argument and instead do the following:

``` python
config = [ {'module': model.linear2, 'sparsity_level': 0.5}]
sparsifier = MySparsifier(config)
```

Commit:
Update on "[sparse] Sparsifier class"


Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"




Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"




Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"




Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"




Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"




Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,182 @@
+            local_args.update(module_config)
+            module = local_args['module']
+            module_path = _module_to_fqn(self.model, module)
+            if module_path and module_path[0] == '.':
```
Can we handle this logic in _module_to_fqn?

z-a-f replied:
I don't think with the current implementation it is that simple -- the name having an empty root name is only resolved when the _module_to_fqn exits all the recursion calls. To make it part of it, I would need to create additional function(wrapper) around the _module_to_fqn.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,182 @@
+        """
+        for config in self.module_groups:
+            module = config['module']
+            if getattr(module, 'mask', None) is None:
```
Can we have masks be only attr of FakeSparsity? If we change FakeSparsity's init to only take the weight shape, this is simpler. Currently there are two masks: one as an attr of the module and one as part of the FakeSparsityModule.
Suggest instead doing:
``` python
for config in self.module_groups:
            module = config['module']
            param = config.get('parametrization', FakeSparsity)
            parametrize.register_parametrization(module, 'weight',
                                                 param(getattr(module,'weight').shape))
```

z-a-f replied:
Oh the intention was that "if there is a mask in the module, it would use that, otherwise it would create one

raghuramank100 replied:
Lets make it explicit, the module should not have a mask, this way mask is always part of fakesparse module.

z-a-f commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,182 @@
+            module = config['module']
+            parametrize.remove_parametrizations(module, 'weight',
+                                                leave_parametrized=True)
+            if getattr(module._parameters, 'mask', None):
```
Because this is a generalized approach, we must consider a case where the mask is part of the module. I will add a check if there is an attribute in the module

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,182 @@
+        format_string += ')'
+        return format_string
+
+    def state_dict(self):
```
This is how I think of the state:
```python
# state returned by getstate
 __getstate__(self):
  return {
            'defaults': self.defaults,
            'state': self.state,
            'module_groups': self.module_groups,
        }
# state_dict
 # {'fqn0': {'name1': Tensor1, 'name2': Tensor2, ....},
     'fqn1': {'name1':Tensor3, 'name2': Tensor4,...}
   }
 # This gets assigned to self.state

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

raghuramank100 commented on test/ao/sparsity/test_sparsifier.py:
```diff
@@ -0,0 +1,57 @@
+        sparsifier0 = ImplementedSparsifier(test=3)
+        sparsifier0.prepare(model, [model.linear])
+        state_dict = sparsifier0.state_dict()
+        sparsifier1 = ImplementedSparsifier()
```
Can we add an explicit check for what the keys and values of the state_dict should look like here?

raghuramank100 commented on torch/ao/sparsity/_mappings.py:
```diff
@@ -0,0 +1,25 @@
+
+def get_static_sparse_quantized_mapping():
+    _static_sparse_quantized_mapping = dict({
+        # Dense
```
Why are the dense mappings present here? Lets make this simpler: We only map from sparse.Linear to the quantized Linear modules. If someone wants to override this, they can.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,213 @@
+            state[original_weight]['mask'] = mask
+            state[original_weight]['fqn'] = g['fqn']
+            # Get all the tensors inside the module_group
+            state[original_weight].update(
```
This doesnt look right. We cannot serialize the original weight as a key to the state.

z-a-f replied:
That what the optimizer does though -- I am mimicking what they do

z-a-f replied:
But I think using fqns as keys should be fine.

raghuramank100 commented on torch/ao/sparsity/sparsifier.py:
```diff
@@ -0,0 +1,213 @@
+                    parametrize.register_parametrization(layer, 'weight', FakeSparsity(p))
+
+            group['module'] = layer
+        self.__setstate__({'module_groups': module_groups})
```
What about restoring the 'state' ?

z-a-f replied:
Forgot about it :/

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

raghuramank100 reviewed (approved):
Please verify tests with latest changes prior to landing

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

Commit:
Update on "[sparsity] Sparsifier class"


Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

[ghstack-poisoned]

z-a-f commented:
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).
